collection_check_boxによるグループ作成の実装

Koh0214 opened the pull request:
# WHAT
・group, user, group_userのモデル、テーブルの関係を作り直した
・htmlのformからrailsのform_forへの書き換え
・collection_check_boxの実装
・グループ名、group_usersの一括保存

# WHY
・グループ機能実装のために必要なため

Commit:
create pull request

Commit:
htmlへのcollection check boxの設置

参考
http://qiita.com/kz_kazuki/items/56957ff4e40a4aa8fd0e

Commit:
check_boxを使っていたので、collection_check_boxへ書き換え

参考
http://qiita.com/hamichamp/items/8cf980e6d5ca9fd7e96f

chat_box_controllerのcreateで情報を見るために、
binding.pryを追加

Commit:
ストロングパラメーターを設定

Commit:
ストロングパラメーターのメソッド名を変更

Koh0214 commented:
binding.pryによるパラメーター許可の確認
![2017-01-12 13 59 44](https://cloud.githubusercontent.com/assets/22772196/21877608/8290fd54-d8cf-11e6-9b56-675f2c417e77.png)

Koh0214 commented:
チャットグループの作成画面
![2017-01-12 14 02 50](https://cloud.githubusercontent.com/assets/22772196/21877658/da67993e-d8cf-11e6-962c-53931c7f5e5c.png)

Commit:
Groupのnameだけ入れられた。GroupUserはエラー。

配列の周りの処理がいけないのかなと思う。

Commit:
form_forのために@groupを用意

Koh0214 commented:
ターミナルで、変数の確認と
データベースへの保存の確認（groups :nameのみ）

![2017-01-12 18 39 08](https://cloud.githubusercontent.com/assets/22772196/21884445/b49576d2-d8f6-11e6-8b6a-ffe51ef57fe1.png)
![2017-01-12 18 40 15](https://cloud.githubusercontent.com/assets/22772196/21884449/b69a0600-d8f6-11e6-9cf2-1e4801550ea4.png)

koheiiwamura commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -1,8 +1,21 @@
 	def create
+		set_params_group
+		set_params_chat_group
+		binding.pry
```
binding.pry消しましょう。

Koh0214 replied:
すみません！！！
消しました！

koheiiwamura commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -1,8 +1,21 @@
+	def set_params_chat_group
+		params.require(:chat_group).permit(:name, :user_ids => [])
+		Group.create(name: params[:chat_group][:name])
+	end
```
このふたつ似たようなメソッドがあるのですが、これはひとつのメソッドだけでグループが作られてユーザーも保存することができます。

Koh0214 replied:
paramsの中身が書き換わってしまうかと勘違いしていました。

保存できることを確認したのちに統合しました！

Commit:
set_params一つに統合、binding.pryの削除

koheiiwamura commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -1,21 +1,19 @@
-	end
-	def set_params_chat_group
 		params.require(:chat_group).permit(:name, :user_ids => [])
 		Group.create(name: params[:chat_group][:name])
```
16行目でGroupをcreateしているのですが、この引数のとり方だと14,15行目が必要ないですね。
createアクションのなかで、Group.new(set_params)にして、それに適したset_paramsメソッドを作ってください。
おそらく事前学習でもやってます。

Koh0214 replied:
createを上に移して、中でset_paramsを読むように修正しました！

Commit:
ストロングパラメータ周りの修正

下でpermitされているのは
:name, :user_idsなどなど。

なので[:chat_group][:name]などの二重シンボルだとnilになる。

KoichiOzaki commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -1,8 +1,18 @@
+
+	private
+	def set_params
+		params.require(:group).permit(:group_users => [])
```
rubyでreturnを省略した場合は最終評価値が返り値となります。
このままでは14行目のコード実行の意味がなくなってしまいますので、15行目で生成しているハッシュにmergeメソッドで結合してあげましょう。

Koh0214 replied:
mergeで統合しました！

Commit:
ストロングパラメータのmergeによる統合

koheiiwamura commented on app/controllers/chat_groups_controller.rb:
```diff
@@ -1,8 +1,17 @@
 	end
 
 	def create
+		Group.create(name: set_params[:name])
```
いい感じになってますね！
create(set_params)だけで生成することができるので、そのように設定してみましょう。

koheiiwamura replied:
mooovi,profyでも似たような処理がされているので、参考にしてみてください。

Commit:
form_forへの書き換え（collection_check_box以外）

Commit:
chat_groupsコントローラーから、groups_controllerに全てを移動。

Commit:
chat_groupsのコントローラー、ビュー、ルートの記述の削除

Commit:
ビューの微調整

Commit:
group_usersをreferenceを使って書き直し

Commit:
group_userのモデルをreferencesを加えて作り直した。

Commit:
references設定をしていなかったgroup_usersモデルを削除、作り直し

Commit:
collection_check_boxesの=を-へ書き換え。

こいつのせいで、2日詰まった。。。。。

Commit:
validationをコメントアウト

Commit:
グループコントローラーのいらない部分を削除

Koh0214 commented:
![2017-01-13 20 30 58](https://cloud.githubusercontent.com/assets/22772196/21928807/4c79062e-d9d0-11e6-8699-2b3ab372a8f2.png)
![2017-01-13 20 31 11](https://cloud.githubusercontent.com/assets/22772196/21928809/4c7ed0d6-d9d0-11e6-8941-de844879245d.png)
![2017-01-13 20 31 19](https://cloud.githubusercontent.com/assets/22772196/21928808/4c795e44-d9d0-11e6-9811-5ec61744403d.png)

Commit:
test_helperの削除

Koh0214 commented:
「チャットメンバーを追加」のテキストフィールドは、現在使わないのでひとまず非表示にしています。

その影響でレイアウトが少し崩れています。

koheiiwamura commented:
LGTM
